File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("RoomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    document,
      querySelector('input[data-behavior="room_speaker"]'),
      addEventListener('keypress', (event) => { // メッセージが入力された時にイベント発火
        if (event.key === 'Enter') { // Enterキーを押下時、speakメソッドを呼び出す
          this.speak(event.target.value);
          event.target.value = '';
          return event.preventDefault();
        }
      });
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    // 受け取ったメッセージをHTML上に追加
    const element = document.querySelector('#messages')
    element.insertAdjacentHTML('beforeend', data['message'])

  },

  speak: function() {
    return this.perform('speak');
  }
});

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    # stream_from "some_channel"
    # ブロードキャスト用のストリーム名を設定
    stream_from "room_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak
    # クライアントがメッセージを送信した時に実行される処理
    message = Message.create!(content: data["message"])
    ActionCable.server.broadcast(
      "room_channel", { message: render_message(message) }
    )
  end

  private

  def render_message
    ApplicationController.render(
      partial: "messages/message",
      locals: { message: message }
    )
  end
end
